token.php

```diff
@@ -0,0 +1,56 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Generate a web service token if a valid user private key is provided.
+ *
+ * @package    mod_applaunch
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__ . '/../../config.php');
+
+$token = required_param('token', PARAM_TEXT); // Course Module ID.
+
+$output = [
+    'wstoken' => '',
+    'errors' => '',
+    'baseurl' => (new moodle_url('/'))->out(),
+    'activityslug' => '',
+];
+
+try {
+    // Check if the token is valid.
+    $userkey = \mod_applaunch\external\ws_token::get_user_key($token);
+    if (!\mod_applaunch\external\ws_token::is_user_key_valid($userkey)) {
+        $output['errors'] .= get_string('error:invalidtoken', 'applaunch') . "\n";
+    }
+
+    // Create user token and populate data.
+    $validuntil = time() + 86400; // Seconds. One day. // TODO: Make this configurable?
+    $output['wstoken'] = \mod_applaunch\external\ws_token::generate_ws_token($userkey, $validuntil);
+    $output['activityslug'] = '/mod/applaunch/view.php?id=' . $userkey->get_instance();
+
```

Comment:
wondering if we should call external_log_token_request like the mobile app launch does

Reply:
This function isn't in totara 12

Reply:
ok easy